single-pool/js/packages/client/src/instructions.ts

```diff
@@ -0,0 +1,327 @@
+      ReadonlyAccount<typeof SYSVAR_CLOCK_ID>,
+      ReadonlyAccount<typeof SYSVAR_STAKE_HISTORY_ID>,
+      ReadonlyAccount<typeof STAKE_CONFIG_ID>,
+      ReadonlyAccount<typeof SYSTEM_PROGRAM_ID>,
+      ReadonlyAccount<typeof TOKEN_PROGRAM_ID>,
+      ReadonlyAccount<typeof STAKE_PROGRAM_ID>,
+    ]
+  > &
+  IInstructionWithData<Uint8Array>;
+
+type DepositStakeInstruction = IInstruction<typeof SINGLE_POOL_PROGRAM_ID> &
+  IInstructionWithAccounts<
+    [
+      ReadonlyAccount<PoolAddress>,
+      WritableAccount<PoolStakeAddress>,
+      WritableAccount<PoolMintAddress>,
+      ReadonlyAccount<PoolStakeAuthorityAddress>,
+      ReadonlyAccount<PoolMintAuthorityAddress>,
+      WritableAccount<Base58EncodedAddress>, // user stake
+      WritableAccount<Base58EncodedAddress>, // user token
+      WritableAccount<Base58EncodedAddress>, // user lamport
+      ReadonlyAccount<typeof SYSVAR_CLOCK_ID>,
+      ReadonlyAccount<typeof SYSVAR_STAKE_HISTORY_ID>,
+      ReadonlyAccount<typeof TOKEN_PROGRAM_ID>,
+      ReadonlyAccount<typeof STAKE_PROGRAM_ID>,
+    ]
+  > &
+  IInstructionWithData<Uint8Array>;
+
+type WithdrawStakeInstruction = IInstruction<typeof SINGLE_POOL_PROGRAM_ID> &
+  IInstructionWithAccounts<
+    [
+      ReadonlyAccount<PoolAddress>,
+      WritableAccount<PoolStakeAddress>,
+      WritableAccount<PoolMintAddress>,
+      ReadonlyAccount<PoolStakeAuthorityAddress>,
+      ReadonlyAccount<PoolMintAuthorityAddress>,
+      WritableAccount<Base58EncodedAddress>, // user stake
+      WritableAccount<Base58EncodedAddress>, // user token
+      ReadonlyAccount<typeof SYSVAR_CLOCK_ID>,
+      ReadonlyAccount<typeof TOKEN_PROGRAM_ID>,
+      ReadonlyAccount<typeof STAKE_PROGRAM_ID>,
+    ]
+  > &
+  IInstructionWithData<Uint8Array>;
+
+type CreateTokenMetadataInstruction = IInstruction<typeof SINGLE_POOL_PROGRAM_ID> &
+  IInstructionWithAccounts<
+    [
+      ReadonlyAccount<PoolAddress>,
+      ReadonlyAccount<PoolMintAddress>,
+      ReadonlyAccount<PoolMintAuthorityAddress>,
+      ReadonlyAccount<PoolMplAuthorityAddress>,
+      WritableSignerAccount<Base58EncodedAddress>, // mpl payer
+      WritableAccount<Base58EncodedAddress>, // mpl account
+      ReadonlyAccount<typeof MPL_METADATA_PROGRAM_ID>,
+      ReadonlyAccount<typeof SYSTEM_PROGRAM_ID>,
+    ]
+  > &
+  IInstructionWithData<Uint8Array>;
+
+type UpdateTokenMetadataInstruction = IInstruction<typeof SINGLE_POOL_PROGRAM_ID> &
+  IInstructionWithAccounts<
+    [
+      ReadonlyAccount<VoteAccountAddress>,
+      ReadonlyAccount<PoolAddress>,
+      ReadonlyAccount<PoolMplAuthorityAddress>,
+      ReadonlySignerAccount<Base58EncodedAddress>, // authorized withdrawer
+      WritableAccount<Base58EncodedAddress>, // mpl account
+      ReadonlyAccount<typeof MPL_METADATA_PROGRAM_ID>,
+    ]
+  > &
+  IInstructionWithData<Uint8Array>;
+
+enum SinglePoolInstructionType {
+  InitializePool,
+  DepositStake,
+  WithdrawStake,
+  CreateTokenMetadata,
+  UpdateTokenMetadata,
+}
```

Comment:
I wonder if it's important to actually write `InitializePool = 0` to make it clear that if you come along and reorder these you're gonna have a bad time (or deprecate one and then remove it).

Also, if you use a `const enum` then the compiler can inline the ones you use and discard the rest: ([demo](https://www.typescriptlang.org/play?#code/MYewdgzgLgBApmArgWxgZQJZgOYBs4AKIIuAkpFAE6LBQbgAqAngA5wwDeAUDLzORjoBDXBgBehYrhgBeGAAYANDz4AROCxARBaKEIDW7OQEZlfGAHVBACwAmlIQHddBozABMZvgGFKcIVBwDCCGYACycHq2AUKyMADMXrwAqizRgcGhEVExcQAsygC+XFxgcI4wyVhQABwAgpQOTAAUANqYOPhEJOTQ1LT0YMxsAHS+-hkhCNlC6UIAugCUQA))